New beatmap information wedge design implementation

mk56-spn opened the pull request:
Another step towards : https://github.com/ppy/osu/issues/18890 being done.

Adds the updated wedge design ( exclusively as a test scene component for now, given it removes a large amount of information, which will be implemented in its own component, as such implementing it directly is unfeasible)

In the file I marked down two issues that are out of my control to fix, namely:

- The mismatched shadow type on the text ( to my knowledge diffused text shadows are not possible currently, given the text is a sprite instead of an SVG and there is no pre rendered diffused shadow i think)).

- And the fact that i have as of yet not heard back from Flyte on how to source the colours for the updated wedge background gradient ( https://www.figma.com/file/DXKwqZhD5yyb1igc3mKo1P?node-id=2980:3361#340801912 )

As a final note I plan to do the implementation of the playcount count present in the design separately if possible, due to it seeming to have more implementation detail than I'd expected.

![updated2](https://user-images.githubusercontent.com/74463310/211616419-fef52121-5de9-498a-b48b-30de8abb537a.png)

mk56-spn committed:
Remove info labels, since they are no longer present in this component in thew new design

mk56-spn committed:
Move stardifficulty logic to main BeatmapInfoWedge class instead of text subclass ( for use by background star rating colour bar )

mk56-spn committed:
Add a test scene and move colour bar to back and adjust positioning of it

mk56-spn committed:
Add coloured star counter

mk56-spn committed:
Simplify passing data from BeatmapInfoWedgeV2.cs to subclass wedgeinfotext

mk56-spn committed:
Remove nullability disabling

mk56-spn committed:
small tweaks, container edge - effect addition..

mk56-spn committed:
Add TODO for text margin const, added pertinent comments to known "issues"

mk56-spn committed:
Remove nullable disable in test.

mk56-spn committed:
Fix margins on top right aligned elements.

bdach commented:
A few basic things from just a visual inspection:

- The stars on the right don't seem to match - design has them centred vertically (and seems to round down to full stars), current implementation has them stuck to top
- As the background in the test is black, there's no way to even tell how this design will work with actual backgrounds
- The designs don't seem to have map status anywhere - you seem to have added them in an educated guess and it seems maybe fine, but input from designer-san would be appreciated if possible

mk56-spn commented:
> A few basic things from just a visual inspection:
> 
>     * The stars on the right don't seem to match - design has them centred vertically (and seems to round down to full stars), current implementation has them stuck to top
> 
>     * As the background in the test is black, there's no way to even tell how this design will work with actual backgrounds
> 
>     * The designs don't seem to have map status anywhere - you seem to have added them in an educated guess and it seems maybe fine, but input from designer-san would be appreciated if possible

1: this is due to it being the old star counter design which  has small stars going up to 10, instead of the new one which cuts off wherever it needs to, old design was left in after frenzy mentioned not being a big fan https://discord.com/channels/188630481301012481/188630652340404224/1049958910720737310 till further discussion is had.

2: black background isnt ideal i agree. will look into

3: yep, but i asked ppy on discord and he said if something is already there that isnt in the drafts it should generally be left there.  https://discord.com/channels/188630481301012481/188630652340404224/1050002698885398528

Joehuu commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,341 @@
+                    RelativeSizeAxes = Axes.Y,
+
+                    // By limiting the width we avoid this box showing up as an outline around the drawables that are on top of it.
+                    Width = 40,
```
If we want to know what value this comes from (figma says 30 plus the corner radius workaround):
`Width = colour_bar_width + corner_radius,` or similiar variable name.

Joehuu commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,341 @@
+                    Origin = Anchor.Centre,
+                    Scale = new Vector2(0.35f),
+                    Shear = -wedged_container_shear,
+                    X = -15,
```
`X = -colour_bar_width / 2,`

Joehuu commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,341 @@
+                {
+                    Masking = true,
+                    // We offset this by the portion of the colour bar underneath we wish to show
+                    X = -30,
```
`X = -colour_bar_width,`

Joehuu commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,341 @@
+                        Name = "Top-left aligned metadata",
+                        Direction = FillDirection.Vertical,
+                        Position = new Vector2(text_margin + shear_width, 12),
+                        Width = .7f,
```
This is quite losing on the extra space:
![image](https://user-images.githubusercontent.com/35318437/212605014-78e1a8e2-e874-470c-9f1b-94ff505c9b2b.png)

And why is it not using padding as before? The width is relative (just placed it in-game for demo):
![osu!_qWdqIy3USp](https://user-images.githubusercontent.com/35318437/212605345-9dc84e60-4f18-4a8e-b6d0-2f6a98661c26.gif)

Joehuu commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,341 @@
+            {
+                base.LoadComplete();
+
+                starRatingDisplay.DisplayedStars.BindValueChanged(s =>
+                {
+                    wedge.starCounter.Current = (float)s.NewValue;
+                    wedge.starCounter.Colour = s.NewValue >= 6.5 ? colours.Orange1 : Colour4.Black.Opacity(0.75f);
+
+                    wedge.difficultyColourBar.FadeColour(colours.ForStarDifficulty(s.NewValue));
+                }, true);
```
Not sure this is worth caching wedge for.

peppy replied:
Definitely feels back-to-front. I'd expose this value and then consume it from the wedge.

mk56-spn committed:
Add back null beatmap test

mk56-spn commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,184 @@
+            AddAssert("check artist", () => infoWedge.Info!.ArtistLabel.Current.Value == $"{ruleset.ShortName}Artist");
+        }
+
+        [SetUpSteps]
+        public void SetUpSteps()
+        {
+            AddStep("reset mods", () => SelectedMods.SetDefault());
+        }
+
+        [Test]
+        public void TestTruncation()
```
@ bdach does this having a background cover  #2 from this https://github.com/ppy/osu/pull/22116#issuecomment-1377760150 suffficiently?

mk56-spn committed:
Address issues that joehuu brought up

mk56-spn committed:
Expose star difficulty to wedge to allow updating starcounter and background colour internally.

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -183,6 +185,14 @@ void removeOldInfo()
                     removeOldInfo();
                     Add(DisplayedContent = loaded);
+
+                    Info.StarRatingDisplay.DisplayedStars.BindValueChanged(s =>
```
Rather than exposing this whole thing, expose just the bindable.

Inside `WedgeInfoText`:

```public Bindable<whatever> DisplayedStars => Info.StarRatingDisplay.DisplayedStars`

mk56-spn replied:
right, thanks ,will do this going forward for similar cases

mk56-spn committed:
Buffer wedge content to avoid opacity issues when showing / hiding

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -202,13 +208,15 @@ public partial class WedgeInfoText : Container
             public OsuSpriteText ArtistLabel { get; private set; } = null!;
 
-            public StarRatingDisplay StarRatingDisplay = null!;
+            private StarRatingDisplay starRatingDisplay = null!;
```
These changes should have been in their own commit. They don't match the commit message.

mk56-spn replied:
my bad, it was an oversight

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,346 @@
+            CornerRadius = corner_radius;
+
+            // We want to buffer the wedge to avoid weird transparency overlaps between the colour bar and the background.
+            Child = bufferedContent = new BufferedContainer
```
No, you really do not want to buffer the whole thing. Have you run the tests and looked at the quality of things in general? Using a buffered container on things like text causes texel misalignment and makes everything more blurry. There's a reason on the original info wedge it was only applied to the background.

You can work around this by using 

```suggestion
            Child = bufferedContent = new BufferedContainer(pixelSnapping: true)
```

but note that if the wedge is ever moved around on the screen using a transform, this will cause it to look less smooth.

Keep in mind there's an inherent overhead with every usage of `BufferedContainer`, too.

peppy replied:
@mk56-spn did you miss this feedback?

mk56-spn replied:
> @mk56-spn did you miss this feedback?

Nope. Working on a fix. Been a tad busy. All should he addressed and done within the day - 2days max

mk56-spn replied:
fixed, i do see a path forward which does not require pixel snapping by separating the text and background into container only one of which is nested in the buffered container, i however did not opt for this since i felt that the pixel snapping look good here ( i tried it on the carousel to get a feel for the different applications and it look god awful because it can be moved by the user and it looked choppy as hell ) and benefited the wedge anyhow by making the line between the ``colourbar`` and the background less muddy

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,346 @@
+
+        private const double animation_duration = 800;
+
+        protected override void PopIn()
+        {
+            this.MoveToX(0, animation_duration, Easing.OutQuint);
+            this.RotateTo(0, animation_duration, Easing.OutQuint);
+            this.FadeIn(transition_duration);
+        }
+
+        protected override void PopOut()
+        {
+            this.MoveToX(-100, animation_duration, Easing.In);
+            this.RotateTo(10, animation_duration, Easing.In);
+            this.FadeOut(transition_duration * 2, Easing.In);
+        }
+
```
This will need reconsideration. The animation was made by me for the previous version, and does not look correct with this one.

mk56-spn replied:
i considered this more of a change to be made during integration, since it'd be easier to see what would fit. but sure ill address

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,346 @@
+            }
+        }
+
+        public override bool IsPresent => base.IsPresent || DisplayedContent == null; // Visibility is updated in the LoadComponentAsync callback
```
What?

This needs reconsideration (yes I know this is in the original code, but it cannot be blindly copied into a newer version).

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,346 @@
+
+            void perform()
+            {
+                void removeOldInfo()
```
double nested local methods? put at the end of the parent method and don't nest this deep.

This needs reconsideration (yes I know this is in the original code, but it cannot be blindly copied into a newer version).

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,346 @@
+        {
+            Scheduler.AddOnce(perform);
+
+            void perform()
```
Just use inline lambda..

```csharp
Scheduler.AddOnce(() => {
...
});
```

mk56-spn replied:
yeah i was having too much faith in existing code, addressed

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,346 @@
+
+            float starAngle = (float)(Math.Atan(shear_width / wedge_height) * (180 / Math.PI));
+
+            // Applying the rotation directly to the StarCounter distorts the stars, hence it is applied to the child container
```
This cannot.

peppy replied:
and this? if you are unsure how to solve, please say so and I will look into a solution.

mk56-spn replied:
> and this? if you are unsure how to solve, please say so and I will look into a solution.


I am unsure but mainly I've been a bit busy for a couple of days. A hint would Definitely be appreciated though

peppy replied:
Put it this way: as a general rule, a drawable should not be interacting with the children of a nested drawable.

The most raw solution would be to fix rotation to work correctly (inside the `StarCounter`). Or make a new derived / new version of the star counter.

mk56-spn replied:
should be fixed now

peppy reviewed (commented):
prelim review there's a lot wrong

mk56-spn committed:
Update ```BeatmapInfoWedgeV2.cs``` animation to be similar to exit transition in ```SongSelect.cs``

mk56-spn committed:
de-nest ```removeOldInfo()```

mk56-spn committed:
Use inline lambda for scheduling

mk56-spn committed:
Fix ```BeatmapInfoWedgeV2.cs``` starCounter needing janky rotation application

mk56-spn committed:
enable pixelSnapping for the ```BufferedContainer``` in BeatmapInfoWedgeV2.cs

mk56-spn committed:
Remove ```IsPresent``` usages

peppy commented:
Something looks broken:

https://user-images.githubusercontent.com/191335/216918774-8a1bf607-afb0-43b1-ab01-1ebaf680bce3.mp4

peppy commented:
Is there a reason there's no background in any of the tests? It seems a bit counter-productive to have the main visual element of the thing being tested missing.

Also the tests being copied are *really* dated, and half are missing. I think doing a pass of the tests to bring them up to scratch would be good (splitting into individual test methods, fixing `show` doing nothing, etc.)

mk56-spn commented:
> Is there a reason there's no background in any of the tests? It seems a bit counter-productive to have the main visual element of the thing being tested missing.
> 
> Also the tests being copied are _really_ dated, and half are missing. I think doing a pass of the tests to bring them up to scratch would be good (splitting into individual test methods, fixing `show` doing nothing, etc.)

the last test has a background but yes i'll try improving them

mk56-spn committed:
Separate wedge visibility test into its own method

mk56-spn committed:
Separate ruleset changing tests into their own method.

Add small clarification for edge colouring in visibility test

mk56-spn committed:
Clean up formatting and wedge placement in testscene a tad,

mk56-spn commented on osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,214 @@
+        [Test]
+        public void TestWedgeVisibility()
+        {
+            // Mostly just in case someone runs this test before others,
```
Not entirely sure if this is OK to do? it alters the visuals somewhat, is a background behind the wedge preferable for showcasing the shadow

peppy replied:
Yes, change the background instead.

peppy replied:
I have no idea why this was resolved. It shouldn't be here. It affects the test PERMANENTLY (you are never recreating the drawables).

mk56-spn committed:
Improve visibility of wedge shading in test scene and fix an issue with excessive roundness on said shadow.

peppy commented:
Why is the star animation delayed?

https://user-images.githubusercontent.com/191335/225549002-520e8832-fa1d-4791-8555-9966aae7302b.mp4


Why is there no way to view the wedge with actual beatmap backgrounds in the test? As with the other components you're working on, I think this is under-tested even by you because there's no way to view the expected game behaviour.

peppy committed:
Adjust shadow to look better

peppy commented on osu.Game/Screens/Select/BeatmapInfoWedgeV2.cs:
```diff
@@ -0,0 +1,339 @@
+                {
+                    Masking = true,
+                    // We offset this by the portion of the colour bar underneath we wish to show
+                    X = -colour_bar_width,
```
This kinda stuff is SO dodgy I don't even know how it's working.

![osu Game Tests 2023-03-16 at 07 57 19](https://user-images.githubusercontent.com/191335/225551447-0b211a54-b784-44d3-9eb6-7a11fb76fd23.png)

I dunno.

mk56-spn replied:
Super late reply. Irl stuff has had me busy, what exactly about it is the dodgy as hell?

Joehuu replied:
I believe this should use right padding instead of a negative X offset?

peppy replied:
Something liek that yes

Joehuu committed:
Merge branch 'master' into beatmap_wedge_clean_ii

Joehuu committed:
Fix truncating sprite text usage

Joehuu committed:
Use right padding instead of negative x offset

Joehuu committed:
Move negative corner radius margin to constructor

Joehuu committed:
Move drawable and binding logic to standard places

Joehuu committed:
Simplify text initialization

Joehuu committed:
Fix wedge showing abruptly in test

Joehuu committed:
Remove weird red edge effect visibility

Joehuu committed:
Fix star rating rolling counter regression

Joehuu committed:
Fix star counter animating weird and delayed

Joehuu committed:
Rename null beatmap test to indicate there's a background

Joehuu commented:
Have pushed changes to address most of the problems + code quality.

Here's a WIP branch with the wedge in-game (to test with how it behaves, not taking designs into account yet): https://github.com/ppy/osu/compare/master...Joehuu:osu:wedge-redesign

peppy committed:
Remove buffered container (doesn't seem like it's doing anything anymore)

peppy committed:
Fix test scene not updating all star counters

peppy commented:
This is looking pretty decent from a code perspective.

One remaining hope I have is to move the star rating out of the reloadable content so it doesn't awkwardly flash on beatmap change:

https://github.com/ppy/osu/assets/191335/5c51a7e1-c385-4e53-8110-8c44467a0fb2

Maybe the rank status can also be moved out (although it would require some consideration as to its animation, but that can come later.

peppy reviewed (changes requested):
As commented

Joehuu committed:
Remove buffered container comment

While it is true that there is a transparency overlap, it is only seen when entering/exiting song select. And there are other components game-wide that exhibit the same issue when transparency changes. Should be reconsidered another time.

Joehuu commented:
Like this?

https://github.com/ppy/osu/assets/35318437/cdae2f33-88b8-45d5-bccb-f7980111211c

Moving the star rating display really improves code quality, so pushing the changes anyway soon.

Joehuu committed:
Remove mods dependency

BPM is not in the wedge anymore.

Joehuu committed:
Make beatmap non-nullable

Doesn't seem like it'll be null after first set: https://github.com/ppy/osu/blob/5454d1caa1927428fe339e3d5b47ddfe45f9dfb6/osu.Game/OsuGameBase.cs#L363

Joehuu committed:
Move actual star updating to star rating display

Joehuu committed:
Move star rating display and status pill out of reloadable content

Joehuu committed:
Fix left padding adding shear width

peppy commented:
Yep, that looks perfect.
